07_errors/runnerdave/map_store.go

```diff
@@ -0,0 +1,52 @@
+package main
+
+// MapStore represents a simple map storage for the Puppy store
+type MapStore struct {
+	puppies map[uint16]Puppy
+}
+
+// NewMapStore creates a new in-memory store with map intialised
+func NewMapStore() *MapStore {
+	return &MapStore{puppies: map[uint16]Puppy{}}
+}
+
+// CreatePuppy saves new puppy if not in store, if it is already returns error
+func (m *MapStore) CreatePuppy(p *Puppy) error {
+	if err := validateValue(p.Value); err != nil {
+		return err
+	}
+	if _, ok := m.puppies[p.ID]; ok {
+		return Errorf(ErrUnknown, "puppy with id %d already exists", p.ID)
+	}
+	m.puppies[p.ID] = *p
+	return nil
+}
+
+// ReadPuppy reads store by Puppy ID
+func (m *MapStore) ReadPuppy(id uint16) (Puppy, error) {
+	if puppy, ok := m.puppies[id]; ok {
+		return puppy, nil
+	}
+	return Puppy{}, Errorf(ErrIDNotFound, "puppy with ID:%d not found", id)
+}
+
+// UpdatePuppy updates puppy with new value if ID present otherwise error
+func (m *MapStore) UpdatePuppy(id uint16, p *Puppy) error {
+	if err := validateValue(p.Value); err != nil {
+		return err
+	}
+	if _, ok := m.puppies[id]; !ok {
+		return Errorf(ErrIDNotFound, "puppy with ID:%d not found", id)
+	}
+	m.puppies[id] = *p
+	return nil
+}
+
+// DeletePuppy deletes a puppy by id from the store
+func (m *MapStore) DeletePuppy(id uint16) error {
+	if _, ok := m.puppies[id]; ok {
```

Comment:
Usual Go style would be to test for `!ok` and return an error, leaving the non-error case at the left edge of the function: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

In `ReadPuppy` you couldn't really do that because of the scope of the `puppy` variable was the `if` block, but in this case you are ignoring that variable.

But this function is small enough that it doesn't really matter, so don't worry about changing it unless you really want to. Just keep it in mind in future - if it could go either way, keep the happy path at the left edge.